src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
I am suggesting this API changes:

- Support only `compare_exchange*` and deprecate `compare_and_swap*`.  They are too similar to each other and we want to avoid method bloat.

    + Maybe we can rename `compare_exchange` into `compare_and_swap`.  I don't have a strong opinion on this, though.

- Change the signature of `compare_exchange*`.  Now it accepts `order: Into<Orderings>` rather than `success: Ordering, failure: Ordering`.  Orderings is effectively a pair of success & failure orderings:

            pub fn compare_exchange<'a>(&self, old: Option<Shared<T>>, new: Option<Shared<T>>,
                                            order: Orderings, guard: &'a Guard)
                                            -> Result<Option<Shared<'a, T>>, Option<Shared<'a, T>>>

- Implement an implicit conversion from Ordering to Orderings using `strongest_failure_ordering`.

I believe in this way, we can avoid method bloat, and support an ergonomic & generic API at the same time.  What do you think?